toolchain/check/testdata/interop/cpp/function/operators.carbon

```diff
@@ -38,33 +38,102 @@ fn F() {
 }
 
 // ============================================================================
-// AddWith
+// Binary operators
 // ============================================================================
 
-// --- add_with.h
+// --- binary_operators.h
 
 class C {};
-auto operator+(C lhs, C rhs) -> C;
 
-// --- import_add_with.carbon
+// Arithmetic.
+auto operator+(C lhs, C rhs) -> C;
+auto operator-(C lhs, C rhs) -> C;
+auto operator*(C lhs, C rhs) -> C;
+auto operator/(C lhs, C rhs) -> C;
+auto operator%(C lhs, C rhs) -> C;
+
+// Bitwise.
+auto operator&(C lhs, C rhs) -> C;
+auto operator|(C lhs, C rhs) -> C;
+auto operator^(C lhs, C rhs) -> C;
+auto operator<<(C lhs, int num_bits) -> C;
+auto operator>>(C lhs, int num_bits) -> C;
+
+// Compound Assignment Arithmetic.
+// TODO: Change `lhs` and return value to reference when it is supported.
+auto operator+=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+auto operator-=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+auto operator*=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+auto operator/=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+auto operator%=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+
+// Compound Assignment Bitwuse.
+// TODO: Change `lhs` and return value to reference when it is supported.
+auto operator&=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+auto operator|=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+auto operator^=(C* _Nonnull lhs, C rhs) -> C* _Nonnull;
+// TODO: Add <<= and >>= when references are supported.
+
+// Relational.
+auto operator==(C lhs, C rhs) -> bool;
+auto operator!=(C lhs, C rhs) -> bool;
+auto operator<(C lhs, C rhs) -> bool;
+auto operator>(C lhs, C rhs) -> bool;
+auto operator<=(C lhs, C rhs) -> bool;
+auto operator>=(C lhs, C rhs) -> bool;
+
+// --- import_binary_operators.carbon
 
 library "[[@TEST_NAME]]";
 
-import Cpp library "add_with.h";
+import Cpp library "binary_operators.h";
 
 fn F() {
   //@dump-sem-ir-begin
-  let c1: Cpp.C = Cpp.C.C();
-  let c2: Cpp.C = Cpp.C.C();
-  let c3: Cpp.C = c1 + c2;
+  var c1: Cpp.C = Cpp.C.C();
+  var c2: Cpp.C = Cpp.C.C();
+
+  // Arithmetic.
+  let addition: Cpp.C = c1 + c2;
+  let subtraction: Cpp.C = c1 - c2;
+  let multiplication: Cpp.C = c1 * c2;
+  let division: Cpp.C = c1 / c2;
+  let modulo: Cpp.C = c1 % c2;
+
+  // Bitwise.
+  let bitwise_and: Cpp.C = c1 & c2;
+  let bitwise_or: Cpp.C = c1 | c2;
+  let bitwise_xor: Cpp.C = c1 ^ c2;
+  let left_shift: Cpp.C = c1 << 3;
+  let right_shift: Cpp.C = c1 >> 5;
+
+  // Compound Assignment Arithmetic.
+  &c1 += c2;
+  &c1 -= c2;
+  &c1 *= c2;
+  &c1 /= c2;
+  &c1 %= c2;
```

Comment:
This is testing pointer arithmetic and I am not sure we want to use Clang to do pointer arithmetic, like we would for calling operator+ on a C++ object itself. This feels more like integer arithmetic which we would want to follow the Carbon rules whatever they might be (like we might disallow this in strict mode or something, and we might want less UB than C++ here almost certainly). Can you confirm that we want to allow C++ to perform pointer arithmetic? If not, then we should be testing pointer arithmetic of `&C` both for `+=` and `+` etc but ensuring we stay on the Carbon side for it.

Reply:
This is an overloaded `operator+=`, not the builtin `+=` on pointers, so we don't actually know this is doing pointer arithmetic -- it's just calling an opaque `operator+=(C* _Nonnull lhs, C rhs)`.

That said, is it really important / relevant that the left-hand side is an operator for this test? Could we test an `operator+=(C lhs, C rhs)` instead, rather than using a test that looks like it might be doing pointer arithmetic?

Reply:
Removed the pointers.
The reason I used pointers is because while doing a full end to end test (see the demo in the description) I noticed we couldn't actually use these operators without pointers since we can't do something like x = (x += y) in Carbon. For the operator calls to actually make effect pointers seemed like a (hacky) workaround that works.
Anyways, for the tests it probably doesn't matter much.